File: tests/CCSharp.Tests/LuaClassTest.cs
using System;
using CCSharp.Attributes;

namespace CCSharp.Tests;

[LuaClass]
public class Person
{
    public string Name { get; }

    public Person(string firstName)
    {
        Name = firstName;
    }
        
    public Person(string firstName, string lastName)
    {
        Name = firstName + " " + lastName;
    }

    public void Greet(string greeting)
    {
        Console.WriteLine(greeting + " " + Name + "!");
    }
}
[LuaProgram]
public static class LuaClassTest
{
    [LuaProperty("uniqueIdentifier")]
    public static Person Person { get; set; }

    [LuaMain]
    public static void Start()
    {
        Person = new Person("gh");
        Person bob = new("Bob");
        bob.Greet("Hi");
        Person boberson = new("Bob", "Boberson");
        boberson.Greet("Hola");
    }
}

File: tests/CCSharp.Tests/lua/LuaClassTest.lua
Person = ({})
Person.__index = Person
Person._CSharpTypes = ({"Person", "Object"})
function Person.new(firstName, _initializer)
 local self = setmetatable(({}), Person);
 self.Name = firstName
 if _initializer then
  for _initializerKey,_initializerValue in pairs(_initializer) do
   self[_initializerKey] = _initializerValue
  end
 end
 return self;
end
function Person.new2(firstName, lastName, _initializer)
 local self = setmetatable(({}), Person);
 self.Name = (firstName.." ")..lastName
 if _initializer then
  for _initializerKey,_initializerValue in pairs(_initializer) do
   self[_initializerKey] = _initializerValue
  end
 end
 return self;
end
function Person:Greet(greeting)
 print(((greeting.." ")..self.Name).."!")
end

function Start()
 uniqueIdentifier = Person.new("gh")
 local person = Person.new("Bob");
 person:Greet("Hi")
 local person2 = Person.new2("Bob", "Boberson");
 person2:Greet("Hola")
end
Start()